### PR TITLE
550 bob test mode

### LINKED ISSUE #550: Add test mode to Bob
In test mode `bob` should be able to start without any configuration. All parameters should be automatically set in the code to a valid default values, cluster config should also be automatically produced in code and consist of one node with one vdisk. Configs should not be stored on disk!
This mode will be useful if someone wants to test Bob but doesn't want to deal with all the configuration files and parameters.

Example: `bobd testmode`

Additionally, several command line arguments should be supported in this mode:
1. `--data /opt/bob` - optional, specifies the path to the directory with bob data
2. `--grpc-port 20000` - optional, changes the port for gRPC API
3. `--restapi-port 8000` - optional, changes the port for REST API
4. `--nodes 10.5.5.1:20000,10.5.5.2:20000` - optional; when specified, `Bob` cluster config automatically generated with all specified nodes containing single physical disk with single vdisk (no replication, just sharding)

Logs in this mode are only written to stdout with `error` level.

After `bobd` started, it should print to stdout information about that plus gRPC API address and REST API urls to perform `put` and `get` requests. Something like:
```
Bob started
Data directory: /opt/bob
gRPC API available at: 10.5.5.1:20000
REST API available at: http://10.5.5.1:8000
REST API Put and Get available at: http://10.5.5.1:8000/data
```

### PR DESCRIPTION
**Pull request by lostl1ght**
Closes #550

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,7 @@ Bob versions changelog
 
 ## [Unreleased]
 #### Added
+-- Bobd test mode
```
Comment (ikopylov) [inline CHANGELOG.md:7]: One dash + issue number
```suggestion
- Bobd test mode (#550)
```

File: bob-apps/bin/bobd.rs [hunk 2]
```diff
@@ -250,6 +387,33 @@ fn check_folders(node: &NodeConfig, init_flag: bool) {
 
 fn get_matches<'a>() -> ArgMatches<'a> {
     let ver = format!("{}\n{}", crate_version!(), BuildInfo::default());
+    let testmode_sc = SubCommand::with_name("testmode")
+        .about("Bob's test mode")
+        .arg(
+            Arg::with_name("data")
+            .help("path to bob data")
```
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:394]: `path to bob data directory`

File: bob-apps/bin/bobd.rs [hunk 3]
```diff
@@ -250,6 +387,33 @@ fn check_folders(node: &NodeConfig, init_flag: bool) {
...
+        )
+        .arg(
+            Arg::with_name("restapi-port")
+            .help("REST API port")
+            .takes_value(true)
+            .long("restapi-port")
+        )
+        .arg(
+            Arg::with_name("nodes")
+            .help("node addresses")
```
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:412]: `Comma separated node addresses. Example: 127.0.0.1:20000,127.0.0.1:20001`

File: bob-apps/bin/bobd.rs [hunk 4]
```diff
@@ -120,6 +148,115 @@ async fn main() {
...
+    let trigger = SizeTrigger::new(_20mb);
+    let roller = FixedWindowRoller::builder()
+        .build( "./log/archive/info.{}.log", 10).unwrap();
+    let policy = CompoundPolicy::new(Box::new(trigger), Box::new(roller));
+    let requests_error = RollingFileAppender::builder()
+        .encoder(Box::new(PatternEncoder::new("{d(%Y-%m-%d %H:%M:%S):<20} {M:>20.30}:{L:>3} {l} {m}{n}")))
+        .build( "./log/error.log", Box::new(policy)).unwrap();
+
+    let config = Config::builder()
+        .appender(Appender::builder().build("stdout", Box::new(stdout)))
```
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:223]: Only `stdout` logger needed for test mode

File: bob-apps/bin/bobd.rs [hunk 5]
```diff
@@ -120,6 +148,115 @@ async fn main() {
...
+        this_node = Some(0);
+        addresses.push(format!("127.0.0.1:{}", port))
+    }
+    let cluster = ClusterConfig::get_testmode(
+        sub_matches.value_of("data").unwrap_or("data").to_string(),
+        addresses).unwrap();
+    let http_api_port = sub_matches.value_of("restapi-port").and_then(|v|  Some(v.parse().unwrap()));
+    let node = cluster.get_testmode_node(this_node.unwrap(), http_api_port).unwrap();
+
+    init_testmode_logger(log::LevelFilter::Info);
```
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:192]: Error level should be used for test mode

File: bob-apps/bin/bobd.rs [hunk 6]
```diff
@@ -120,6 +148,115 @@ async fn main() {
...
+                .appender("stdout")
+                .appender("requests_error")
+                .additive(false)
+                .build("bob_grpc", loglevel))
+        .logger(Logger::builder()
+                .appender("stdout")
+                .appender("requests_error")
+                .additive(false)
+                .build("pearl", loglevel))
+        .build(Root::builder().appender("stdout").build(loglevel))
```
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:255]: Maybe it would be enough to create only a Root logger? Can you check if logs will be written if only the root logger is created?

File: bob-apps/bin/bobd.rs [hunk 7]
```diff
@@ -17,34 +17,62 @@ use std::fs::create_dir;
...
-    }
-
-    if matches.value_of("node").is_none() {
-        eprintln!("Expect node config");
-        eprintln!("use --help");
-        return;
-    }
+    let cluster;
+    let node;
+    if let (_, Some(sub_matches)) = matches.subcommand() {
```
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:40]: You should check subcommand name

File: bob-apps/bin/bobd.rs [hunk 8]
```diff
@@ -17,34 +17,62 @@ use std::fs::create_dir;
...
+        match configure_testmode(sub_matches) {
+            Ok((c, n)) => {
+                cluster = c;
+                node = n;
+            }
+            Err(e) => {
+                eprintln!("{}", e);
```
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:47]: `Initialization error: {}`
`use --help`

File: bob-apps/bin/bobd.rs [hunk 9]
```diff
@@ -120,6 +148,115 @@ async fn main() {
     }
 }
 
+fn configure_testmode(sub_matches: &ArgMatches) -> Result<(ClusterConfig, NodeConfig), String> {
+    let mut addresses = Vec::with_capacity(1);
+    let port = match sub_matches.value_of("grpc-port") {
+        Some(v) => v.parse().unwrap(),
```
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:154]: `unwrap` will panic. You should add `anyhow` context and return Error with `?` operator

File: bob-apps/bin/bobd.rs [hunk 10]
```diff
@@ -120,6 +148,115 @@ async fn main() {
...
+    if let Some(node_list) = sub_matches.value_of("nodes") {
+        let available_ips: Vec<String> = NetworkInterface::show().unwrap().into_iter().filter_map(|itf|
+            match itf.addr.unwrap() {
+                Addr::V4(addr) => {
+                    Some(addr.ip.to_string())
+                },
+                _ => None
+        }).collect();
+
+        for (addr, index) in node_list.split(",").zip(0..usize::MAX) {
```
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:167]: `enumerate` instead of `zip`

File: bob-apps/bin/bobd.rs [hunk 11]
```diff
@@ -120,6 +148,115 @@ async fn main() {
...
+            let split = &addr.split_once(":").unwrap();
```
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:168]: Again: return contextual error to the caller

File: bob-apps/bin/bobd.rs [hunk 12]
```diff
@@ -120,6 +148,115 @@ async fn main() {
...
+            if this_node.is_none() {
+                for ip in available_ips.iter() {
+                    if ip == split.0 && port == split.1.parse::<u16>().unwrap() {
```
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:171]: Again

File: bob-apps/bin/bobd.rs [hunk 13]
```diff
@@ -120,6 +148,115 @@ async fn main() {
...
+                Addr::V4(addr) => {
+                    Some(addr.ip.to_string())
+                },
+                _ => None
+        }).collect();
+
+        for (addr, index) in node_list.split(",").zip(0..usize::MAX) {
+            let split = &addr.split_once(":").unwrap();
+            if this_node.is_none() {
+                for ip in available_ips.iter() {
```
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:170]: `filter().map().next()` probably be shorter

File: bob-apps/bin/bobd.rs [hunk 14]
```diff
@@ -120,6 +148,115 @@ async fn main() {
...
+        }
+        if this_node.is_none() {
+            return Err(String::from("No available IPs found"));
+        }
+    } else {
+        this_node = Some(0);
+        addresses.push(format!("127.0.0.1:{}", port))
+    }
+    let cluster = ClusterConfig::get_testmode(
+        sub_matches.value_of("data").unwrap_or("data").to_string(),
```
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:187]: Maybe it would be good to add `this_node` index at the end of the name to avoid collisions if several nodes starts from single binary

File: bob-apps/bin/bobd.rs [hunk 15]
```diff
@@ -120,6 +148,115 @@ async fn main() {
...
+        addresses).unwrap();
+    let http_api_port = sub_matches.value_of("restapi-port").and_then(|v|  Some(v.parse().unwrap()));
```
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:189]: Return contextual error instead of `unwrap`

File: bob-apps/bin/bobd.rs [hunk 16]
```diff
@@ -120,6 +148,115 @@ async fn main() {
...
+    let node = cluster.get_testmode_node(this_node.unwrap(), http_api_port).unwrap();
+
+    init_testmode_logger(log::LevelFilter::Info);
+
+    check_folders(&node, true);
+
+    println!("Bob has started");
```
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:196]: Bob actually has not started at this point. I suggest to write `Bob starting`

File: bob-apps/bin/bobd.rs (hunk 15 above)
Comment (lostl1ght) [reply bob-apps/bin/bobd.rs]: Done: [1e08373](https://github.com/qoollo/bob/pull/588/commits/1e0837312a5aa82e7351366ebf65dff582dd9a9c)

File: bob-apps/bin/bobd.rs (hunk 12 above)
Comment (lostl1ght) [reply bob-apps/bin/bobd.rs]: Done: [1e08373](https://github.com/qoollo/bob/pull/588/commits/1e0837312a5aa82e7351366ebf65dff582dd9a9c)

File: bob-apps/bin/bobd.rs (hunk 11 above)
Comment (lostl1ght) [reply bob-apps/bin/bobd.rs]: Done: [1e08373](https://github.com/qoollo/bob/pull/588/commits/1e0837312a5aa82e7351366ebf65dff582dd9a9c)

File: bob-apps/bin/bobd.rs (hunk 10 above)
Comment (lostl1ght) [reply bob-apps/bin/bobd.rs]: Done: [1e08373](https://github.com/qoollo/bob/pull/588/commits/1e0837312a5aa82e7351366ebf65dff582dd9a9c)

File: bob-apps/bin/bobd.rs (hunk 9 above)
Comment (lostl1ght) [reply bob-apps/bin/bobd.rs]: Done: [1e08373](https://github.com/qoollo/bob/pull/588/commits/1e0837312a5aa82e7351366ebf65dff582dd9a9c)

File: bob-apps/bin/bobd.rs (hunk 8 above)
Comment (lostl1ght) [reply bob-apps/bin/bobd.rs]: Done: [1e08373](https://github.com/qoollo/bob/pull/588/commits/1e0837312a5aa82e7351366ebf65dff582dd9a9c)

File: bob-apps/bin/bobd.rs (hunk 5 above)
Comment (lostl1ght) [reply bob-apps/bin/bobd.rs]: Done: [1e08373](https://github.com/qoollo/bob/pull/588/commits/1e0837312a5aa82e7351366ebf65dff582dd9a9c)

File: bob-apps/bin/bobd.rs (hunk 2 above)
Comment (lostl1ght) [reply bob-apps/bin/bobd.rs]: Fix: [cdc7316](https://github.com/qoollo/bob/pull/588/commits/cdc73164e2a79a83876565868f58bcedc4659ebf)

File: bob-apps/bin/bobd.rs (hunk 3 above)
Comment (lostl1ght) [reply bob-apps/bin/bobd.rs]: Fix: [cdc7316](https://github.com/qoollo/bob/pull/588/commits/cdc73164e2a79a83876565868f58bcedc4659ebf)

File: bob-apps/bin/bobd.rs (hunk 7 above)
Comment (lostl1ght) [reply bob-apps/bin/bobd.rs]: Fix: [3bdb33d](https://github.com/qoollo/bob/pull/588/commits/3bdb33d49a6cd64e060e5fc935e35767e4da086e)

File: bob-apps/bin/bobd.rs (hunk 14 above)
Comment (lostl1ght) [reply bob-apps/bin/bobd.rs]: I would also suggest creating a directory if it does not exist. It should probably be handled by `check_folders` function.

File: bob-apps/bin/bobd.rs (hunk 16 above)
Comment (lostl1ght) [reply bob-apps/bin/bobd.rs]: Fix: [f0b0215](https://github.com/qoollo/bob/pull/588/commits/f0b02157ff45022f0859f4807d820eeae9b256c0)

File: bob-apps/bin/bobd.rs (hunk 13 above)
Comment (lostl1ght) [reply bob-apps/bin/bobd.rs]: Not really. However, I've removed parsing on each iteration: [8bd8716](https://github.com/qoollo/bob/pull/588/commits/8bd87160a09b27ed573191ecda5541f393c2c380)

File: bob-apps/bin/bobd.rs (hunk 6 above)
Comment (lostl1ght) [reply bob-apps/bin/bobd.rs:219]: Yes: [0cf1f5f](https://github.com/qoollo/bob/pull/588/commits/0cf1f5f52ef2601db946f55f99307fb75336e6c7)

File: bob-apps/bin/bobd.rs [hunk 17]
```diff
@@ -120,6 +146,81 @@ async fn main() {
...
+        let available_ips: Vec<String> = NetworkInterface::show()?.into_iter().filter_map(|itf|
+            match itf.addr? {
+                Addr::V4(addr) => {
+                    Some(addr.ip.to_string())
+                },
+                _ => None
+        }).collect();
+
+        for (index, addr) in node_list.split(",").enumerate() {
+            let split = &addr.split_once(":").context("could not find address in --nodes")?;
```
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:166]: Change the error text to better reflect the reason

File: bob-apps/bin/bobd.rs [hunk 18]
```diff
@@ -120,6 +146,81 @@ async fn main() {
...
+            let in_ip = split.0;
+            let in_port = split.1.parse::<u16>().context("could not parse port in --nodes")?;
```
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:168]: Change the error text to better reflect the reason

File: bob-apps/bin/bobd.rs [hunk 19]
```diff
@@ -120,6 +146,81 @@ async fn main() {
...
+                        Some(index)
+                    } else {
+                        None
+                    }
+                })
+            }
+            addresses.push(String::from(addr));
+        }
+        if this_node.is_none() {
+            return Err(anyhow!("no available ips found"));
```
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:181]: Change the error text to better reflect the reason

File: bob-apps/bin/bobd.rs [hunk 20]
```diff
@@ -120,6 +146,81 @@ async fn main() {
...
 
+fn configure_testmode(sub_matches: &ArgMatches) -> AnyResult<(ClusterConfig, NodeConfig)> {
+    let mut addresses = Vec::with_capacity(1);
+    let port = match sub_matches.value_of("grpc-port") {
+        Some(v) => v.parse().context("could not parse --grpc-port")?,
+        None => 20000
+    };
+    let mut this_node = None;
+    if let Some(node_list) = sub_matches.value_of("nodes") {
+        let available_ips: Vec<String> = NetworkInterface::show()?.into_iter().filter_map(|itf|
```
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:157]: Use `HashSet<>` instead of `Vec<>` to simplify lookup

File: bob-apps/bin/bobd.rs [hunk 21]
```diff
@@ -120,6 +146,81 @@ async fn main() {
...
+                _ => None
+        }).collect();
+
+        for (index, addr) in node_list.split(",").enumerate() {
+            let split = &addr.split_once(":").context("could not find address in --nodes")?;
+            let in_ip = split.0;
+            let in_port = split.1.parse::<u16>().context("could not parse port in --nodes")?;
+            if this_node.is_none() {
+                this_node = available_ips.iter().find_map(|ip| {
+                    if ip == in_ip && port == in_port {
```
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:171]: `port` comparison does not depend on the collection element and can be performed outside

File: bob-apps/bin/bobd.rs [hunk 22]
```diff
@@ -120,6 +146,81 @@ async fn main() {
...
+                        Some(index)
+                    } else {
+                        None
+                    }
+                })
+            }
+            addresses.push(String::from(addr));
+        }
+        if this_node.is_none() {
```
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:180]: Below you have a line ` let this_node = this_node.unwrap();`. That line is a better place to convert to an error (you can use `context(..)` function)

File: bob-apps/bin/bobd.rs [hunk 23]
```diff
@@ -120,6 +146,81 @@ async fn main() {
...
+
+    init_testmode_logger(log::LevelFilter::Error);
+
+    check_folders(&node, true);
+
+    println!("Bob is starting");
+    let n = &cluster.nodes()[this_node];
+    println!("Data directory: {}", n.disks()[0].path());
+    println!("gRPC API available at: {}", n.address());
+    let ip = node.http_api_address();
```
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:205]: `rest_api_ip`

File: bob-apps/bin/bobd.rs [hunk 24]
```diff
@@ -120,6 +146,81 @@ async fn main() {
...
+    let p = node.http_api_port();
```
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:206]: `rest_api_port`

Comment (ikopylov) [inline bob-apps/bin/bobd.rs:206]:
```suggestion
    let rest_api_port = node.http_api_port();
```

File: bob-apps/bin/bobd.rs (hunk 23 above)
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:205]:
```suggestion
    let res_api_ip = node.http_api_address();
```

File: bob-common/src/configs/cluster.rs [hunk 25]
```diff
@@ -423,6 +423,49 @@ impl Cluster {
             Ok(config)
         }
     }
+
+    pub fn get_testmode(path: String, addresses: Vec<String>) -> AnyResult<Self> {
+        let disks = vec![DiskPath::new("disk_0".to_string(), path)];
+        let len = addresses.len();
```
Comment (ikopylov) [inline bob-common/src/configs/cluster.rs:429]: `len` is not needed as a separate variable

File: bob-common/src/configs/cluster.rs [hunk 26]
```diff
@@ -423,6 +423,49 @@ impl Cluster {
...
+        let mut nodes = Vec::with_capacity(len);
+        let mut vdisk = VDisk::new(0);
+        for (address, i) in addresses.into_iter().zip(0..len) {
```
Comment (ikopylov) [inline bob-common/src/configs/cluster.rs:432]: `enumerate()`

File: bob-common/src/configs/cluster.rs [hunk 27]
```diff
@@ -423,6 +423,49 @@ impl Cluster {
...
+            let node = Node {
+                name: format!("node_{i}"),
+                address,
+                disks: disks.clone()
+            };
+            let replica = Replica::new(node.name().to_string(), node.disks()[0].name().to_string());
+            vdisk.push_replica(replica);
```
Comment (ikopylov) [inline bob-common/src/configs/cluster.rs:439]: In test mode cluster should work with multiple shards and without any replication. You should create new vdisk for every node

File: bob-common/src/configs/cluster.rs [hunk 28]
```diff
@@ -423,6 +423,49 @@ impl Cluster {
...
+
+        if let Err(e) = config.validate() {
+            let msg = format!("config is not valid: {e}");
+            Err(anyhow!(msg))
+        } else {
+            Ok(config)
+        }
+    }
+
+    pub fn get_testmode_node(&self, n_node: usize, rest_port: Option<u16>) -> AnyResult<NodeConfig> {
```
Comment (ikopylov) [inline bob-common/src/configs/cluster.rs:458]:
```suggestion
    pub fn get_testmode_node_config(&self, n_node: usize, rest_port: Option<u16>) -> AnyResult<NodeConfig> {
```

File: bob-common/src/configs/cluster.rs [hunk 29]
```diff
@@ -423,6 +423,49 @@ impl Cluster {
...
+        let node = &self.nodes()[n_node];
```
Comment (ikopylov) [inline bob-common/src/configs/cluster.rs:459]: Need to handle index out of bounds condition. Use `get(n_node).context(...)`

File: bob-common/src/configs/node.rs [hunk 30]
```diff
@@ -439,6 +458,28 @@ impl Pearl {
         }
         unreachable!()
     }
+
+    pub fn get_testmode(alien_disk: &str) -> Self {
+        Self {
+            max_blob_size: Pearl::default_max_blob_size(),
+            max_data_in_blob: Pearl::default_max_data_in_blob(),
```
Comment (ikopylov) [inline bob-common/src/configs/node.rs:465]: Reduce default value to `100_000`

File: bob-common/src/configs/node.rs [hunk 31]
```diff
@@ -736,6 +777,37 @@ impl NodeConfig {
...
         8
     }
+
+    pub fn get_testmode(node_name: &str, disk_name: &str, rest_port: Option<u16>) -> Self {
+        Self {
+             log_config: String::from("dummy"),
+             users_config: String::from("dummy"),
+             name: String::from(node_name),
+             quorum: 1,
+             operation_timeout: String::from("3sec"),
```
Comment (ikopylov) [inline bob-common/src/configs/node.rs:787]:
```suggestion
             operation_timeout: String::from("60sec"),
```

File: bob-common/src/configs/node.rs [hunk 32]
```diff
@@ -736,6 +777,37 @@ impl NodeConfig {
...
+             count_interval: NodeConfig::default_count_interval(),
+             cluster_policy: String::from("quorum"),
+             backend_type: String::from("pearl"),
+             pearl: Some(Pearl::get_testmode(disk_name)),
+             metrics: Some(MetricsConfig::default()), 
+             bind_ref: Arc::default(),
+             disks_ref: Arc::default(),
+             cleanup_interval: String::from("1h"),
+             open_blobs_soft_limit: Some(2),
+             open_blobs_hard_limit: Some(10),
```
Comment (ikopylov) [inline bob-common/src/configs/node.rs:798]:
```suggestion
             open_blobs_hard_limit: None,
```

File: bob-common/src/configs/node.rs [hunk 33]
```diff
@@ -736,6 +777,37 @@ impl NodeConfig {
...
+             check_interval: String::from("5000ms"),
+             count_interval: NodeConfig::default_count_interval(),
+             cluster_policy: String::from("quorum"),
+             backend_type: String::from("pearl"),
+             pearl: Some(Pearl::get_testmode(disk_name)),
+             metrics: Some(MetricsConfig::default()), 
+             bind_ref: Arc::default(),
+             disks_ref: Arc::default(),
+             cleanup_interval: String::from("1h"),
+             open_blobs_soft_limit: Some(2),
```
Comment (ikopylov) [inline bob-common/src/configs/node.rs:797]:
```suggestion
             open_blobs_soft_limit: None,
```

File: bob-apps/Cargo.toml [hunk 34]
```diff
@@ -47,8 +47,9 @@ lazy_static = "1.4"
 log = "0.4"
 log4rs = "1.2"
 metrics = { version = "0.17", features = ["std"] }
+network-interface = "0.1.2"
+prost = "0.9"
```
Comment (ikopylov) [inline bob-apps/Cargo.toml:51]: Incorrect version downgrade. Revert this change and change below (where `prost = "0.11"` was removed)

File: bob-apps/bin/bobd.rs [hunk 35]
```diff
@@ -25,35 +25,65 @@ use std::fs::create_dir;
...
-        err
-    }).expect("Node config parsing error");
+        let node_config_file = matches.value_of("node").unwrap();
+        println!("Node config: {:?}", node_config_file);
+        node = cluster.get(node_config_file).await.map_err(|err| {
+            eprintln!("Node config parsing error: {}", err);
+            err
+        }).expect("Node config parsing error");
+
+        check_folders(&node, matches.is_present("init_folders"));
```
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:85]: `check_folders` is also called below. Looks like the call below should be removed, because `init_folders` flag is only valid here

File: bob-apps/bin/bobd.rs [hunk 36]
```diff
@@ -139,6 +169,74 @@ async fn main() {
...
+                Addr::V4(addr) => {
+                    Some(addr.ip.to_string())
+                },
+                _ => None
+        }).collect();
+
+        for (index, addr) in node_list.split(",").enumerate() {
+            let split = &addr.split_once(":").context("expected nodes address to have format ip:port")?;
+            let in_ip = split.0;
+            let in_port = split.1.parse::<u16>().context("failed to parse port")?;
```
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:191]: Looks like it would be simpler to parse value into [SocketAddrV4](https://doc.rust-lang.org/nightly/core/net/struct.SocketAddrV4.html#). Also for simplicity it would be good to collect `Ipv4Addr` above from `NetworkInterface::show()` (not perform `to_string()` conversion)

File: bob-common/src/configs/node.rs [hunk 37]
```diff
@@ -441,6 +460,30 @@ impl Pearl {
         }
         unreachable!()
     }
+
+    pub fn get_testmode(alien_disk: &DiskName) -> Self {
+        Self {
+            max_blob_size: ByteUnit::GB,
```
Comment (ikopylov) [inline bob-common/src/configs/node.rs:466]: Please, increase also value returned from `default_max_blob_size` to 1Gb

File: bob-apps/bin/bobd.rs [hunk 38]
```diff
@@ -139,6 +167,72 @@ async fn main() {
...
+    if let Some(node_list) = sub_matches.value_of("nodes") {
+        let available_ips: HashSet<_> = NetworkInterface::show()?.into_iter().filter_map(|itf|
+            match itf.addr? {
+                Addr::V4(addr) => {
+                    Some(addr.ip)
+                },
+                _ => None
+        }).collect();
+
+        for (index, addr) in node_list.split(",").enumerate() {
```
Comment (ikopylov) [inline bob-apps/bin/bobd.rs:186]: It would be good to do `trim` on `addr` (`let addr = addr.trim();`)

**Review by ikopylov — APPROVED** [review]
One minor comment. Otherwise, LGTM

**Review by ikopylov — APPROVED** [review]
LGTM
